integration-cli/docker_cli_build_test.go

```diff
@@ -4941,6 +4941,65 @@ func (s *DockerSuite) TestBuildLabelsCache(c *check.C) {
 
 }
 
+func (s *DockerSuite) TestBuildNotVerboseSuccess(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	// This test makes sure that -q works correctly when build is successful:
+	// stdout has only the image ID (long image ID) and stderr is empty.
+	_, stdout, stderr, err := buildImageWithStdoutStderr("quiet_build_success", "FROM busybox", true, "-q", "--force-rm", "--rm")
+	if err != nil {
+		c.Fatal(fmt.Errorf("Test shouldn't fail, but got the following error: %s", err.Error()))
+	}
+	outRegexp := regexp.MustCompile("^(sha256:|)[a-z0-9]{64}\\n$")
+	if outRegexp.Find([]byte(stdout)) == nil {
+		c.Fatalf("Test expected stdout to match the [%v] regexp, but it is [%v]", outRegexp, stdout)
+	}
+	if stderr != "" {
+		c.Fatalf("Test expected stderr to be empty, but it is [%#v]", stderr)
+	}
+}
+
+func (s *DockerSuite) TestBuildNotVerboseFailure(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+	// This test makes sure that -q works correctly when build fails by
+	// comparing between the stderr output in quiet mode and in stdout
+	// and stderr output in verbose mode
+	tt := []struct {
+		TestName  string
+		BuildCmds string
+	}{
+		{"quiet_build_no_from_at_the_beginning", "RUN whoami"},
+		{"quiet_build_unknown_instr", "FROMD busybox"},
+		{"quiet_build_not_exists_image", "FROM busybox11"},
+	}
+
+	for _, te := range tt {
+		_, _, qstderr, qerr := buildImageWithStdoutStderr(te.TestName, te.BuildCmds, false, "-q", "--force-rm", "--rm")
+		_, vstdout, vstderr, verr := buildImageWithStdoutStderr(te.TestName, te.BuildCmds, false, "--force-rm", "--rm")
+		if verr == nil || qerr == nil {
+			c.Fatal(fmt.Errorf("Test [%s] expected to fail but didn't", te.TestName))
+		}
+		if qstderr != vstdout+vstderr {
+			c.Fatal(fmt.Errorf("Test[%s] expected that quiet stderr and verbose stdout are equal; quiet [%v], verbose [%v]", te.TestName, qstderr, vstdout))
+		}
+	}
+}
+
+func (s *DockerSuite) TestBuildNotVerboseFailureRemote(c *check.C) {
```

Comment:
Because of the 4 variants at https://github.com/docker/docker/pull/17428/files#diff-3a1632a75e011091e0bae6f8c8435fd2L93  can we add a test for when the URL points to a github repo and for when the Dockerfile comes from stdin ?

Aside from that I think this is ready to go!